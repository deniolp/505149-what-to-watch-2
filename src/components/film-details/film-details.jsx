import React from 'react';
import PropTypes from 'prop-types';
import {Redirect} from 'react-router-dom';

import Header from '../header/header';
import Tabs from '../tabs/tabs';
import withLabel from '../../hocs/with-label/with-label';
import FilmsList from '../films-list/films-list';
import Footer from '../footer/footer';
import filmsListMock from '../../mocks/films';

const TabsWrapped = withLabel(Tabs);

const FilmDetails = (props) => {
  let {film} = props;

  const id = props.match.params.id;
  if (!film) {
    film = filmsListMock.find((it) => it.id === +id);
    if (!film) {
      return <Redirect to="/"></Redirect>;
    }
  }
  const filteredByGenreFilms = filmsListMock.filter((item) => item.genre === film.genre && item.name !== film.name);

  return <React.Fragment>
    <section className="movie-card movie-card--full">
      <div className="movie-card__hero">
        <div className="movie-card__bg">
          <img src="/img/bg-the-grand-budapest-hotel.jpg" alt={film.name} />
        </div>
        <Header />
        <div className="movie-card__wrap">
          <div className="movie-card__desc">
            <h2 className="movie-card__title">{film.name}</h2>
            <p className="movie-card__meta">
              <span className="movie-card__genre">{film.genre}</span>
              <span className="movie-card__year">{film.year}</span>
            </p>

            <div className="movie-card__buttons">
              <button className="btn btn--play movie-card__button" type="button">
                <svg viewBox="0 0 19 19" width="19" height="19">
                  <use xlinkHref="#play-s"></use>
                </svg>
                <span>Play</span>
              </button>
              <button className="btn btn--list movie-card__button" type="button">
                <svg viewBox="0 0 19 20" width="19" height="20">
                  <use xlinkHref="#add"></use>
                </svg>
                <span>My list</span>
              </button>
              <a href="add-review.html" className="btn movie-card__button">Add review</a>
            </div>
          </div>
        </div>
      </div>

      <div className="movie-card__wrap movie-card__translate-top">
        <div className="movie-card__info">
          <div className="movie-card__poster movie-card__poster--big">
            <img src="/img/the-grand-budapest-hotel-poster.jpg" alt="The Grand Budapest Hotel poster" width="218" height="327" />
          </div>
          <TabsWrapped film={film}/>
        </div>
      </div>
    </section>

    <div className="page-content">
      <section className="catalog catalog--like-this">
        <h2 className="catalog__title">More like this</h2>
        <div className="catalog__movies-list">
          <FilmsList
            films={filteredByGenreFilms}
            filmsCounter={4}
          />
        </div>
      </section>
      <Footer />
    </div></React.Fragment>;
};

FilmDetails.propTypes = {
  film: PropTypes.shape({
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    src: PropTypes.string.isRequired,
    preview: PropTypes.string.isRequired,
    genre: PropTypes.string.isRequired,
    year: PropTypes.number.isRequired,
    score: PropTypes.number.isRequired,
    ratingLevel: PropTypes.string.isRequired,
    ratingCount: PropTypes.number.isRequired,
    duration: PropTypes.number.isRequired,
    description: PropTypes.string.isRequired,
    director: PropTypes.string.isRequired,
    starring: PropTypes.arrayOf(PropTypes.string).isRequired,
    reviews: PropTypes.arrayOf(PropTypes.shape({
      text: PropTypes.string.isRequired,
      author: PropTypes.string.isRequired,
      time: PropTypes.string.isRequired,
      rating: PropTypes.number.isRequired,
    })),
  }),
  match: PropTypes.object.isRequired,
};

export default FilmDetails;
