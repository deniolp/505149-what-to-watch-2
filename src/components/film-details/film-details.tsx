import * as React from 'react';
import {connect} from 'react-redux';
import {Link, Redirect, withRouter} from 'react-router-dom';

import Header from '../header/header';
import Tabs from '../tabs/tabs';
import withLabel from '../../hocs/with-label/with-label';
import FilmsList from '../films-list/films-list';
import Footer from '../footer/footer';
import {Operation} from '../../reducer/reducer';
import {getFilmsWithChangedUrl} from '../../reducer/selectors';
import {Film} from "../../types";

interface Props {
  onOpenCloseVideoButtonClick: (film: Film) => void;
  onLoadFilms: () => void;
  films: Film[];
  isAuthorizationRequired: boolean;
  onPostFavorite: (id: number, isFavorite: boolean, isPromo: boolean) => void;
  history: {
    push: (path: string) => void;
  };
  match: {
    params: {
      id: number;
    };
  };
}

const TabsWrapped = withLabel(Tabs);

const FilmDetails = (props: Props): React.SFC => {
  const {
    onOpenCloseVideoButtonClick,
    onLoadFilms,
    films,
    isAuthorizationRequired,
    onPostFavorite,
    history
  } = props;

  const renderPage = (film, filteredByGenreFilms): React.F => {
    return <React.Fragment>
      <section
        className="movie-card movie-card--full"
        style={{
          backgroundColor: `${film.backgroundColor}`,
        }}>
        <div className="movie-card__hero">
          <div className="movie-card__bg">
            <img src={film.backgroundImage} alt={film.name} />
          </div>
          <Header />
          <div className="movie-card__wrap">
            <div className="movie-card__desc">
              <h2 className="movie-card__title">{film.name}</h2>
              <p className="movie-card__meta">
                <span className="movie-card__genre">{film.genre}</span>
                <span className="movie-card__year">{film.released}</span>
              </p>

              <div className="movie-card__buttons">
                <button
                  className="btn btn--play movie-card__button"
                  type="button"
                  onClick={(): void => onOpenCloseVideoButtonClick(film)}
                >
                  <svg viewBox="0 0 19 19" width="19" height="19">
                    <use xlinkHref="#play-s"></use>
                  </svg>
                  <span>Play</span>
                </button>
                <button
                  className="btn btn--list movie-card__button"
                  type="button"
                  onClick={(): void => isAuthorizationRequired ? history.push(`/login`) : onPostFavorite(film.id, film.isFavorite, false)}
                >
                  {film.isFavorite ? <svg viewBox="0 0 18 14" width="18" height="14">
                    <use xlinkHref="#in-list"></use>
                  </svg> : <svg viewBox="0 0 19 20" width="19" height="20">
                    <use xlinkHref="#add"></use>
                  </svg>}
                  <span>My list</span>
                </button>
                {!isAuthorizationRequired && <Link to={`/film/${film.id}/review`} className="btn movie-card__button">Add review</Link>}
              </div>
            </div>
          </div>
        </div>

        <div className="movie-card__wrap movie-card__translate-top">
          <div className="movie-card__info">
            <div className="movie-card__poster movie-card__poster--big">
              <img src={film.posterImage} alt={film.name} width="218" height="327" />
            </div>
            <TabsWrapped
              film={film}
            />
          </div>
        </div>
      </section>

      <div className="page-content">
        <section className="catalog catalog--like-this">
          <h2 className="catalog__title">More like this</h2>
          <div className="catalog__movies-list">
            <FilmsList
              films={filteredByGenreFilms}
              filmsCounter={4}
            />
          </div>
        </section>
        <Footer />
      </div></React.Fragment>;
  };

  if (films.length !== 0) {
    const id = props.match.params.id;
    const film = films.find((it) => it.id === +id);
    if (!film) {
      return <Redirect to="/"></Redirect>;
    }
    const filteredByGenreFilms = films.filter((item) => item.genre === film.genre && item.name !== film.name);

    return renderPage(film, filteredByGenreFilms);
  } else {
    onLoadFilms();
  }
  return null;
};

const mapStateToProps = (state, ownProps): object => Object.assign({}, ownProps, {
  films: getFilmsWithChangedUrl(state),
});

const mapDispatchToProps = (dispatch) => ({
  onLoadFilms: (): void => dispatch(Operation.loadFilms()),
});

export {FilmDetails};
export default connect(mapStateToProps, mapDispatchToProps)(React.memo(withRouter(FilmDetails)));
