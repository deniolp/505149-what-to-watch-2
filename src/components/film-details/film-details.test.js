import React from 'react';
import renderer from 'react-test-renderer';
import {BrowserRouter} from 'react-router-dom';

import {FilmDetails} from './film-details';
import filmsListMock from '../../mocks/films';

describe(`FilmDetails`, () => {
  it(`renders correctly`, () => {
    window.history.pushState({}, ``, `/film/1`);
    const tree = renderer.create(
        <BrowserRouter><FilmDetails
          films={filmsListMock}
          match={{
            params: {
              id: 1,
            },
          }}
          onOpenCloseVideoButtonClick={jest.fn()}
          onLoadFilms={jest.fn()}
        /></BrowserRouter>).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
