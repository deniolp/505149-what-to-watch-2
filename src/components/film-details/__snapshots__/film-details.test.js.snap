// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmDetails renders correctly 1`] = `
Array [
  <section
    className="movie-card movie-card--full"
  >
    <div
      className="movie-card__hero"
    >
      <div
        className="movie-card__bg"
      >
        <img
          alt="Aviator"
          src="/img/bg-the-grand-budapest-hotel.jpg"
        />
      </div>
      <h1
        className="visually-hidden"
      >
        WTW
      </h1>
      <header
        className="page-header movie-card__head"
      >
        <div
          className="logo"
        >
          <a
            className="logo__link"
            href="/"
          >
            <span
              className="logo__letter logo__letter--1"
            >
              W
            </span>
            <span
              className="logo__letter logo__letter--2"
            >
              T
            </span>
            <span
              className="logo__letter logo__letter--3"
            >
              W
            </span>
          </a>
        </div>
        <div
          className="user-block"
        >
          <div
            className="user-block__avatar"
          >
            <img
              alt="User avatar"
              height="63"
              src="/img/avatar.jpg"
              width="63"
            />
          </div>
        </div>
      </header>
      <div
        className="movie-card__wrap"
      >
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            Aviator
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Drama
            </span>
            <span
              className="movie-card__year"
            >
              2008
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <button
              className="btn btn--play movie-card__button"
              type="button"
            >
              <svg
                height="19"
                viewBox="0 0 19 19"
                width="19"
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              className="btn btn--list movie-card__button"
              type="button"
            >
              <svg
                height="20"
                viewBox="0 0 19 20"
                width="19"
              >
                <use
                  xlinkHref="#add"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
            <a
              className="btn movie-card__button"
              href="add-review.html"
            >
              Add review
            </a>
          </div>
        </div>
      </div>
    </div>
    <div
      className="movie-card__wrap movie-card__translate-top"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster movie-card__poster--big"
        >
          <img
            alt="The Grand Budapest Hotel poster"
            height="327"
            src="/img/the-grand-budapest-hotel-poster.jpg"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <nav
            className="movie-nav movie-card__nav"
          >
            <ul
              className="movie-nav__list"
            >
              <li
                className="movie-nav__item movie-nav__item--active"
              >
                <a
                  className="movie-nav__link"
                  onClick={[Function]}
                  style={
                    Object {
                      "cursor": "pointer",
                    }
                  }
                >
                  Overview
                </a>
              </li>
              <li
                className="movie-nav__item"
              >
                <a
                  className="movie-nav__link"
                  onClick={[Function]}
                  style={
                    Object {
                      "cursor": "pointer",
                    }
                  }
                >
                  Details
                </a>
              </li>
              <li
                className="movie-nav__item"
              >
                <a
                  className="movie-nav__link"
                  onClick={[Function]}
                  style={
                    Object {
                      "cursor": "pointer",
                    }
                  }
                >
                  Reviews
                </a>
              </li>
            </ul>
          </nav>
          <div
            className="movie-rating"
          >
            <div
              className="movie-rating__score"
            >
              8.3
            </div>
            <p
              className="movie-rating__meta"
            >
              <span
                className="movie-rating__level"
              >
                Good
              </span>
              <span
                className="movie-rating__count"
              >
                187
                 ratings
              </span>
            </p>
          </div>
          <div
            className="movie-card__text"
          >
            <p>
              Deserunt ex esse tempor aliqua labore ex ipsum ipsum. Minim elit anim consectetur eu cillum. Quis ut ut minim anim sit irure anim non officia id ex. Elit in ea ad elit consequat magna culpa excepteur Lorem. Excepteur sit sit enim irure enim in amet excepteur. Sunt ullamco et cupidatat est adipisicing occaecat nostrud voluptate aliquip consequat commodo adipisicing. Ipsum exercitation est consectetur aute magna anim dolore id enim.
            </p>
            <p>
                  Sunt ullamco et cupidatat est adipisicing occaecat nostrud voluptate aliquip consequat commodo adipisicing. Ipsum exercitation est consectetur aute magna anim dolore id enim.
            </p>
            <p
              className="movie-card__director"
            >
              <strong>
                Director: 
                Christopher Nolan
              </strong>
            </p>
            <p
              className="movie-card__starring"
            >
              <strong>
                Starring: 
                Bill Murray, Jack Nicholson, Edward Norton, Jude Law, Dustin Hoffman
                 and other
              </strong>
            </p>
          </div>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog catalog--like-this"
    >
      <h2
        className="catalog__title"
      >
        More like this
      </h2>
      <div
        className="catalog__movies-list"
      >
        <article
          className="small-movie-card catalog__movies-card"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <a
            href="/film/2"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
            >
              <video
                height="100%"
                muted={true}
                poster="/img/shutter-island.jpg"
                width="100%"
              >
                <source
                  src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
                  type="video/webm"
                />
              </video>
            </div>
          </a>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="/film/2"
              onClick={[Function]}
            >
              Shutter Island
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <a
            href="/film/3"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
            >
              <video
                height="100%"
                muted={true}
                poster="/img/macbeth.jpg"
                width="100%"
              >
                <source
                  src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
                  type="video/webm"
                />
              </video>
            </div>
          </a>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="/film/3"
              onClick={[Function]}
            >
              Macbeth
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <a
            href="/film/4"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
            >
              <video
                height="100%"
                muted={true}
                poster="/img/revenant.jpg"
                width="100%"
              >
                <source
                  src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
                  type="video/webm"
                />
              </video>
            </div>
          </a>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="/film/4"
              onClick={[Function]}
            >
              Revenant
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <a
            href="/film/5"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
            >
              <video
                height="100%"
                muted={true}
                poster="/img/mindhunter.jpg"
                width="100%"
              >
                <source
                  src="https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm"
                  type="video/webm"
                />
              </video>
            </div>
          </a>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="/film/5"
              onClick={[Function]}
            >
              Mindhunter
            </a>
          </h3>
        </article>
      </div>
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
          href="/"
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          © 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
